### DISCUSSION TITLE
When / how to start the lsp server

### DISCUSSION THREAD

**Discussion by pedz** [emacs-lsp/lsp-mode#3760] [Q&A] [answered]
I'm new to lsp.  I'm using lsp-docker for a specific project which is starting Solargraph inside the container for Ruby files.  Emacs is outside the container.  Currently, that is the only place I'm using lsp-mode.  

What I need (I think) is to do `(lsp t)` the **first** time I start editing a Ruby file inside the project.  I *assume* calling `(lsp t)` every time a Ruby file within the project is loaded would be bad.  Is that correct?

Currently I have an eval in my .dir-locals.el file that calls `(lsp t)`.  The first time I edit a Ruby file, *Messages* has "LSP :: Connected to [rubyls-docker:47989/starting]." but editing subsequent Ruby files in the project has only "LSP :: Connected to [rubyls-docker:47989]." so it appears to be finding the same server.

I'm just wondering if I'm doing things right.  `lsp-mode` is not autoloaded but `lsp` is so simply enabling `lsp-mode` didn't work.

Thank you for your time

**Comment by kiennq** [comment, marked as answer]
You can use `(lsp-deferred)` which will run `(lsp)` when the buffer becomes visible. Also, call them on the `ruby-mode-hook`. That would make sure the lsp will be started every time you open a new Ruby file.
And you don't have to worry about calling `lsp` on every new source file open, it would not have much negative impact on those
